pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -284,71 +178,40 @@ static __always_inline int output_dns_event(struct __sk_buff *skb,
 	// Skip this if skb_val == NULL (gadget_socket_lookup did not set pid_tgid we use in the query key)
 	// or if event->timestamp == 0 (kernels before 5.8 don't support bpf_ktime_get_boot_ns, and the patched
 	// version IG injects always returns zero).
-	if (skb_val != NULL && event->timestamp > 0) {
+	if (skb_val != NULL && event.timestamp > 0) {
+		union dnsflags flags;
+		flags.flags = load_half(skb, DNS_OFF + offsetof(struct dnshdr,
+								flags));
+		id = load_half(skb, DNS_OFF + offsetof(struct dnshdr, id));
+		__u8 qr = flags.qr;
+
 		struct query_key_t query_key = {
 			.pid_tgid = skb_val->pid_tgid,
-			.id = event->id,
+			.id = id,
 		};
-		if (event->qr == DNS_QR_QUERY &&
-		    event->pkt_type == PACKET_OUTGOING) {
+		if (qr == DNS_QR_QUERY && event.pkt_type == PACKET_OUTGOING) {
 			bpf_map_update_elem(&query_map, &query_key,
-					    &event->timestamp, BPF_NOEXIST);
-		} else if (event->qr == DNS_QR_RESP &&
-			   event->pkt_type == PACKET_HOST) {
+					    &event.timestamp, BPF_NOEXIST);
+		} else if (flags.qr == DNS_QR_RESP &&
+			   event.pkt_type == PACKET_HOST) {
 			__u64 *query_ts =
 				bpf_map_lookup_elem(&query_map, &query_key);
 			if (query_ts != NULL) {
 				// query ts should always be less than the event ts, but check anyway to be safe.
-				if (*query_ts < event->timestamp) {
-					event->latency_ns =
-						event->timestamp - *query_ts;
+				if (*query_ts < event.timestamp) {
+					event.latency_ns =
+						event.timestamp - *query_ts;
 				}
 				bpf_map_delete_elem(&query_map, &query_key);
 			}
 		}
 	}
 
-	// size of full structure - addresses + only used addresses
-	unsigned long long size =
-		sizeof(*event) - MAX_ADDR_ANSWERS * 16 + anaddrcount * 16;
-	bpf_perf_event_output(skb, &events, BPF_F_CURRENT_CPU, event, size);
+	__u64 skb_len = skb->len;
+	bpf_perf_event_output(skb, &events, skb_len << 32 | BPF_F_CURRENT_CPU,
```

Comment:
Could you please describe why you passed from `BPF_F_CURRENT_CPU` to `skb_len << 32 | BPF_F_CURRENT_CPU`?

Reply:
This is the way to submit the skb and custom data (the struct). It doesn't seem to be documented in https://man7.org/linux/man-pages/man7/bpf-helpers.7.html, I had to check the implication of [perf_submit_skb()](https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#4-perf_submit_skb) to check how it worked.

Reply:
Oh, that's what I was missing.

Reply:
BTW, [this](https://medium.com/@nurkholish.halim/a-deep-dive-into-ebpf-writing-an-efficient-dns-monitoring-2c9dea92abdf) is an interesting blogpost using the same approach.

Reply:
TIL!

I see bcc handles that automatically:
https://github.com/iovisor/bcc/blob/6cd272189771bbc7120d9d834e2588b93d8e9cfa/src/cc/frontends/clang/b_frontend_action.cc#L1017
```
            "((__u64)" + skb_len + " << 32) | BPF_F_CURRENT_CPU, " +
```